<a href="https://colab.research.google.com/github/ilakshmiteja/Diamonds-Sklearn-pipeline-column-transformer/blob/main/sklearn_pipeline_and_column_transformer_on_diamonds_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing necessary libraries

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import RobustScaler, PowerTransformer, OrdinalEncoder, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

Loading the dataset

In [2]:
diamonds = sns.load_dataset('diamonds')

Viewing the dataset

In [3]:
diamonds

carat        cut color clarity  depth  table  price     x     y     z
0       0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1       0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2       0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3       0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4       0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

[53940 rows x 10 columns]

Checking for missing values and duplicate entries

In [4]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   carat    53940 non-null  float64 
 1   cut      53940 non-null  category
 2   color    53940 non-null  category
 3   clarity  53940 non-null  category
 4   depth    53940 non-null  float64 
 5   table    53940 non-null  float64 
 6   price    53940 non-null  int64   
 7   x        53940 non-null  float64 
 8   y        53940 non-null  float64 
 9   z        53940 non-null  float64 
dtypes: category(3), float64(6), int64(1)
memory usage: 3.0 MB


From above information, we can see that there are no missing values in the dataset.
<br>There are three category columns:
1. cut
2. color
3. clarity
<br>All the remaining columns are numerical columns.
<br>'Price' is our target column.

In [5]:
diamonds.duplicated().sum()

146

We can see that there are 146 duplicates. We will drop the duplicate entries.

In [6]:
diamonds.drop_duplicates(inplace = True)

We will now see the statistical measures of the dataset.

In [8]:
diamonds.describe()

carat         depth         table         price             x  \
count  53794.00000  53794.000000  53794.000000  53794.000000  53794.000000   
mean       0.79778     61.748080     57.458109   3933.065082      5.731214   
std        0.47339      1.429909      2.233679   3988.114460      1.120695   
min        0.20000     43.000000     43.000000    326.000000      0.000000   
25%        0.40000     61.000000     56.000000    951.000000      4.710000   
50%        0.70000     61.800000     57.000000   2401.000000      5.700000   
75%        1.04000     62.500000     59.000000   5326.750000      6.540000   
max        5.01000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53794.000000  53794.000000  
mean       5.734653      3.538714  
std        1.141209      0.705037  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.030000  
max       58.900000     31.800000

We see that there could possibly be outliers in price, x, y and z features. However, these outliers can be caused because of unique or special diamonds.

Splitting Independent features and target feature.

In [9]:
X = diamonds.drop(columns = ['price'])
y = diamonds['price']

Splitting X, y variables to train and test

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 42)

Identifying ordinal, nominal categorical variables and continuous variables

In [12]:
X_Ordinal = ['cut']
X_Nominal = ['color','clarity']
X_cont = ['carat','depth','table','x','y','z']

Ordinal Categories Pipeline

In [16]:
ordinal_pipeline = Pipeline( steps = [
    ('SimpleImputing',SimpleImputer(strategy = 'most_frequent')),
    ('ordinalEncoding',OrdinalEncoder(categories=[['Premium','Very Good','Good','Fair','Ideal']]))
])

Nominal Categories Pipeline

In [17]:
nominal_pipeline = Pipeline(steps = [
    ('SimpleImputing',SimpleImputer(strategy = 'most_frequent')),
    ('oneHotEncoding',OneHotEncoder(sparse_output = False, drop='first'))
])

Continuous variables Pipleline

In [18]:
cont_pipeline = Pipeline(steps = [
    ('SimpleImputing',SimpleImputer(strategy = 'median')),
    ('robustScaling',RobustScaler()),
    ('powerTransformation',PowerTransformer())
])

Column Transformer for wrapping all the continuous and categorical pipelines.

In [19]:
pre_col_trans = ColumnTransformer(transformers = [
    ('ordinalPipeline',ordinal_pipeline,X_Ordinal),
    ('nominalPipeline',nominal_pipeline,X_Nominal),
    ('continuousPipeline',cont_pipeline,X_cont)
    ], remainder = 'passthrough')

Final Pipeline

In [20]:
final_pipeline = Pipeline(steps=[
    ('preColTransformer', pre_col_trans)
    ])

Final Pipeline Diagram

In [25]:
final_pipeline

Pipeline(steps=[('preColTransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('ordinalPipeline',
                                                  Pipeline(steps=[('SimpleImputing',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('ordinalEncoding',
                                                                   OrdinalEncoder(categories=[['Premium',
                                                                                               'Very '
                                                                                               'Good',
                                                                                               'Good',
                                                                                               'Fair',
                                                                                               'Ideal']]))]),
                                                  ['cut']),
                                                 ('nominalPipeline',
                                                  Pipeline(steps=[('SimpleImputing',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('oneHotEncoding',
                                                                   OneHotEncoder(drop='first',
                                                                                 sparse_output=False))]),
                                                  ['color', 'clarity']),
                                                 ('continuousPipeline',
                                                  Pipeline(steps=[('SimpleImputing',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('robustScaling',
                                                                   RobustScaler()),
                                                                  ('powerTransformation',
                                                                   PowerTransformer())]),
                                                  ['carat', 'depth', 'table',
                                                   'x', 'y', 'z'])]))])

Applying fit_transform on our Train data

In [21]:
X_train_ft = pd.DataFrame(final_pipeline.fit_transform(X_train),
                          columns = final_pipeline.get_feature_names_out(),
                          index = X_train.index)

In [22]:
X_train_ft

ordinalPipeline__cut  nominalPipeline__color_E  \
12820                   4.0                       0.0   
19997                   1.0                       0.0   
6099                    0.0                       0.0   
37984                   4.0                       0.0   
24865                   0.0                       0.0   
...                     ...                       ...   
11311                   0.0                       0.0   
44869                   4.0                       0.0   
38271                   4.0                       0.0   
860                     0.0                       0.0   
15832                   0.0                       0.0   

       nominalPipeline__color_F  nominalPipeline__color_G  \
12820                       0.0                       1.0   
19997                       1.0                       0.0   
6099                        0.0                       1.0   
37984                       0.0                       0.0   
24865                       0.0                       1.0   
...                         ...                       ...   
11311                       0.0                       0.0   
44869                       0.0                       1.0   
38271                       0.0                       0.0   
860                         0.0                       0.0   
15832                       0.0                       1.0   

       nominalPipeline__color_H  nominalPipeline__color_I  \
12820                       0.0                       0.0   
19997                       0.0                       0.0   
6099                        0.0                       0.0   
37984                       0.0                       0.0   
24865                       0.0                       0.0   
...                         ...                       ...   
11311                       1.0                       0.0   
44869                       0.0                       0.0   
38271                       0.0                       0.0   
860                         0.0                       0.0   
15832                       0.0                       0.0   

       nominalPipeline__color_J  nominalPipeline__clarity_IF  \
12820                       0.0                          0.0   
19997                       0.0                          0.0   
6099                        0.0                          0.0   
37984                       0.0                          0.0   
24865                       0.0                          0.0   
...                         ...                          ...   
11311                       0.0                          0.0   
44869                       0.0                          0.0   
38271                       1.0                          0.0   
860                         1.0                          0.0   
15832                       0.0                          0.0   

       nominalPipeline__clarity_SI1  nominalPipeline__clarity_SI2  \
12820                           0.0                           0.0   
19997                           0.0                           0.0   
6099                            1.0                           0.0   
37984                           0.0                           0.0   
24865                           0.0                           0.0   
...                             ...                           ...   
11311                           0.0                           0.0   
44869                           1.0                           0.0   
38271                           1.0                           0.0   
860                             1.0                           0.0   
15832                           0.0                           0.0   

       nominalPipeline__clarity_VS1  nominalPipeline__clarity_VS2  \
12820                           0.0                           1.0   
19997                           0.0                           0.0   
6099                            0.0                

Applying transform on our Test data

In [23]:
X_test_ft = pd.DataFrame(final_pipeline.transform(X_test),
                         columns = final_pipeline.get_feature_names_out(),
                         index = X_test.index)

In [24]:
X_test_ft

ordinalPipeline__cut  nominalPipeline__color_E  \
43657                   3.0                       0.0   
4274                    1.0                       1.0   
47412                   4.0                       0.0   
44437                   0.0                       1.0   
13975                   4.0                       0.0   
...                     ...                       ...   
43980                   1.0                       0.0   
1115                    1.0                       0.0   
48829                   4.0                       0.0   
42876                   3.0                       1.0   
52752                   0.0                       0.0   

       nominalPipeline__color_F  nominalPipeline__color_G  \
43657                       0.0                       0.0   
4274                        0.0                       0.0   
47412                       0.0                       1.0   
44437                       0.0                       0.0   
13975                       0.0                       0.0   
...                         ...                       ...   
43980                       0.0                       0.0   
1115                        0.0                       0.0   
48829                       0.0                       1.0   
42876                       0.0                       0.0   
52752                       1.0                       0.0   

       nominalPipeline__color_H  nominalPipeline__color_I  \
43657                       0.0                       0.0   
4274                        0.0                       0.0   
47412                       0.0                       0.0   
44437                       0.0                       0.0   
13975                       0.0                       1.0   
...                         ...                       ...   
43980                       1.0                       0.0   
1115                        1.0                       0.0   
48829                       0.0                       0.0   
42876                       0.0                       0.0   
52752                       0.0                       0.0   

       nominalPipeline__color_J  nominalPipeline__clarity_IF  \
43657                       1.0                          0.0   
4274                        0.0                          0.0   
47412                       0.0                          0.0   
44437                       0.0                          0.0   
13975                       0.0                          0.0   
...                         ...                          ...   
43980                       0.0                          0.0   
1115                        0.0                          0.0   
48829                       0.0                          0.0   
42876                       0.0                          0.0   
52752                       0.0                          0.0   

       nominalPipeline__clarity_SI1  nominalPipeline__clarity_SI2  \
43657                           0.0                           0.0   
4274                            0.0                           1.0   
47412                           0.0                           0.0   
44437                           0.0                           0.0   
13975                           1.0                           0.0   
...                             ...                           ...   
43980                           0.0                           1.0   
1115                            0.0                           0.0   
48829                           0.0                           0.0   
42876                           1.0                           0.0   
52752                           0.0                           0.0   

       nominalPipeline__clarity_VS1  nominalPipeline__clarity_VS2  \
43657                           0.0                           1.0   
4274                            0.0                           0.0   
47412                           1.0                